pkg/pgmodel/sql_ingest.go

```diff
@@ -539,38 +542,115 @@ func runInserter(conn pgxConn, in chan copyRequest) {
+
+		// sort to prevent deadlocks
+		sort.Slice(insertBatch, func(i, j int) bool {
+			return insertBatch[i].table < insertBatch[j].table
+		})
+
+		//merge tables to prevent deadlocks (on row order) and for efficiency
+		for i := 0; i < len(insertBatch); i++ {
+			//for loop since may need to merge multiple table
+			for i < len(insertBatch)-2 && insertBatch[i].table == insertBatch[i+1].table {
+				insertBatch[i].data.absorb(insertBatch[i+1].data)
+				insertBatch[i+1].data.release()
+				//delete element i+1
+				copy(insertBatch[i+1:], insertBatch[i+2:])
+				insertBatch[len(insertBatch)-1] = copyRequest{}
+				insertBatch = insertBatch[:len(insertBatch)-1]
+			}
+		}
+
+		doInsertOrFallback(conn, insertBatch...)
+		for i := range insertBatch {
+			insertBatch[i] = copyRequest{}
+		}
+		insertBatch = insertBatch[:0]
+	}
+}
+
+var getBatchMutex = &sync.Mutex{}
+
+func inserterGetBatch(batch []copyRequest, in chan copyRequest) ([]copyRequest, bool) {
+	//This mutex is not for safety, but rather for better batching.
+	//It guarantees that only one copier is reading from the channel at one time
+	//This ensures bigger batches as well as less spread of a
+	//single http request to multiple copiers, decreasing latency via less sync
+	//overhead especially on low-pressure systems.
+	getBatchMutex.Lock()
+	defer getBatchMutex.Unlock()
+	req, ok := <-in
+	if !ok {
+		return batch, false
+	}
+	batch = append(batch, req)
+
+	//we use a small timeout to prevent low-pressure systems from using up too many
+	//txns and putting pressure on system
+	timeout := time.After(20 * time.Millisecond)
+hot_gather:
+	for len(batch) < cap(batch) {
+		select {
+		case r2 := <-in:
+			batch = append(batch, r2)
+		case <-timeout:
+			break hot_gather
+		}
+	}
+	return batch, true
+}
+
+func doInsertOrFallback(conn pgxConn, reqs ...copyRequest) {
+	err := doInsert(conn, reqs...)
+	if err != nil {
+		insertBatchErrorFallback(conn, reqs...)
+		return
+	}
+
+	for i := range reqs {
+		reqs[i].data.reportResults(nil)
+		reqs[i].data.release()
+	}
+}
+
+func insertBatchErrorFallback(conn pgxConn, reqs ...copyRequest) {
+	for i := range reqs {
+		reqs[i].data.batch.ResetPosition()
+		err := doInsert(conn, reqs[i])
 		if err != nil {
-			err = insertErrorFallback(conn, req, err)
+			err = insertErrorFallback(conn, err, reqs[i])
 		}
 
-		req.data.reportResults(err)
-		pendingBuffers.Put(req.data)
+		reqs[i].data.reportResults(err)
+		reqs[i].data.release()
 	}
 }
 
 // certain errors are recoverable, handle those we can
 //   1. if the table is compressed, decompress and retry the insertion
-func insertErrorFallback(conn pgxConn, req copyRequest, err error) error {
-	err = tryRecovery(conn, req, err)
-	if err != nil {
-		log.Warn("msg", fmt.Sprintf("time out while processing error for %s", req.table), "err", err.Error())
-		return err
+func insertErrorFallback(conn pgxConn, err error, reqs ...copyRequest) error {
+	for _, req := range reqs {
+		err = tryRecovery(conn, err, req)
```

Comment:
Just verifying: do we test that calling `decompressChunks()` succeeds when the chunks aren't compressed?

Reply:
this comment may still be relevant